### PR TITLE
UI/heart-health

### PR DESCRIPTION
**Pull request by nriedman**
# Display Vitals History in Heart Health Page

## :recycle: Current situation & Problem
Currently, the VitalsManager collects the user's health data from Firestore and stores it in several arrays. However, as of now we only use this information to display the user's most recent vitals. The Heart Health page will allow the user to interact with the entirety of their health data in graph and list form.

## :gear: Release Notes 
- Extended the VitalsManager to include Symptom Scores History (from the bi-weekly KCCQ surveys)
- Implemented UI components for displaying the user's health data in a graph, with available time resolutions of daily for the past 30 days, weekly for the past 3 months, and monthly for the past 6 months.
- Multiple measurements in the same time interval are displayed as averages for that time interval.
- On tap or swipe of the graphs, a UI component appears with details of the selected time interval. When no point is selected, an average across the entire date range is displayed in the top leading edge.
- Includes a list of all the patient's measurements for the chosen vital, with the option to delete for weight, heart rate, and blood pressure (delete disabled for symptom scores).

|   |   |   |   |
| --- | --- | --- | --- |
| IMG_4950 | IMG_4951 | IMG_4952 | IMG_4953 |
|  IMG_4954 | IMG_4955 | IMG_4956 | IMG_4957 |
| IMG_4958 | IMG_4959 | IMG_4960 | IMG_4961 |
| IMG_4962 | IMG_4963 | IMG_4964 | IMG_4965 |

## :books: Documentation
See inline documentation.

## :white_check_mark: Testing
Includes thorough Unit and UI Testing for all major features such as data point aggregation, handling empty vitals data, and deleting measurements.

### Code of Conduct & Contributing Guidelines 

By submitting creating this pull request, you agree to follow our [Code of Conduct](https://github.com/StanfordBDHG/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/StanfordBDHG/.github/blob/main/CONTRIBUTING.md):
- [X] I agree to follow the [Code of Conduct](https://github.com/StanfordBDHG/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/StanfordBDHG/.github/blob/main/CONTRIBUTING.md).

### CODE REVIEW

**Comment by nriedman** [comment]
@PSchmiedmayer, by the way, the picker that we'd like to turn into a title is defined in SymptomsPicker.swift, and is used in the main content view defined in SymptomsContentView.swift.

**Review by PSchmiedmayer — COMMENTED** [review]
@nriedman Thank you for the work here; this is really taking shape, great job! 🚀

@pauljohanneskraft Feel free to provide a first review once @nriedman has finished up the latest improvements in the PR.

File: ENGAGEHF/HeartHealth/Vitals/VitalsContentView.swift [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+        case .bloodPressure:
+            guard doubleValues.count == 2 else {
+                return "ERROR"
+            }
+            
+            let systolic = Int(doubleValues[0])
+            let diastolic = Int(doubleValues[1])
+            
+            return "\(systolic)/\(diastolic)"
+        }
```
[the remark this answers is not in the document]
Comment (pauljohanneskraft) [reply ENGAGEHF/HeartHealth/Vitals/VitalsContentView.swift:90]: We might also want to use a different alternative string than `ERROR` - how about an empty string or "-"?

File: ENGAGEHF/ReusableElements/DisplayMeasurement.swift [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+// SPDX-License-Identifier: MIT
+//
+
+import SwiftUI
+
+
+struct DisplayMeasurement: View {
+    var quantity: String
+    var units: String
+    var type: String
```
Comment (pauljohanneskraft) [inline ENGAGEHF/ReusableElements/DisplayMeasurement.swift:15]:
```suggestion
    let quantity: String
    let units: String
    let type: String
```

File: ENGAGEHF/SharedContext/FHIRSystem.swift [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+enum FHIRSystem {
+    case loinc
+    
+    
+    var url: URL {
+        switch self {
+        case .loinc: URL(string: "http://loinc.org")!
+        }
+    }
+}
```
Comment (pauljohanneskraft) [inline ENGAGEHF/SharedContext/FHIRSystem.swift:21]:
```suggestion
enum FHIRSystem {
    // swiftlint-disable-next force-unwrapping
    static let loinc = URL(string: "http://loinc.org")!
}
```

What do you think about this instead?

File: ENGAGEHF/VitalsManager/Extensions/HKSample+Graphable.swift [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+import ExceptionCatcher
+import Foundation
+import SpeziHealthKit
+
+
+/// Conformance to the Graphable protocol allows HKSamples to be displayed no matter what concrete subclass the instance is
+extension HKSample: Graphable {
+    /// Returns the date associated with the measurement
+    /// For now, assumes each sample is associated with a single point in time, not a period
+    public var date: Date { startDate }
```
Comment (pauljohanneskraft) [inline ENGAGEHF/VitalsManager/Extensions/HKSample+Graphable.swift:18]: Is there a reason, why this is needed rather than simply using the startDate wherever we need a single date? Might be confusing for some types of HKSample that are not bound to a single time instant

File: ENGAGEHF/VitalsManager/Extensions/HKSample+Graphable.swift [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+    
+    
+    /// Returns the Double value corresponding to each quantity in the Sample in the given units, if compatable
+    /// For now, only supports HKQuantitySamples and HKCorrelations
+    public func getDoubleValues(for unitString: String) -> [Double] {
```
Comment (pauljohanneskraft) [inline ENGAGEHF/VitalsManager/Extensions/HKSample+Graphable.swift:23]: Is there a way to never have to do this conversion from string to HKUnit in the first place?

File: ENGAGEHF/VitalsManager/Extensions/SymptomScore+Graphable.swift [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+//
+// SPDX-License-Identifier: MIT
+//
+
+import Foundation
+
+
+extension SymptomScore: Graphable {
+    public func getDoubleValues(for symtomString: String) -> [Double] {
+        guard let symptomType = SymptomsType(rawValue: symtomString) else {
```
Comment (pauljohanneskraft) [inline ENGAGEHF/VitalsManager/Extensions/SymptomScore+Graphable.swift:14]:
```suggestion
    public func getDoubleValues(for symptomString: String) -> [Double] {
        guard let symptomType = SymptomsType(rawValue: symptomString) else {
```

File: ENGAGEHF/VitalsManager/SymptomScore.swift [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+public struct SymptomScore: Identifiable, Equatable, Codable {
+    @DocumentID public var id: String?
+    public var date: Date
+    public var overallScore: Double
+    public var physicalLimitsScore: Double
+    public var socialLimitsScore: Double
+    public var qualityOfLifeScore: Double
+    public var specificSymptomsScore: Double
+    public var dizzinessScore: Double
+}
```
Comment (pauljohanneskraft) [inline ENGAGEHF/VitalsManager/SymptomScore.swift:25]:
```suggestion
/// The score representing the result of a patient's response to a KCCQ survey
/// Parameters are specified in compliance with:
/// https://github.com/StanfordBDHG/ENGAGE-HF-Firebase/tree/web-data-scheme
public struct SymptomScores: Identifiable, Equatable, Codable {
    @DocumentID public var id: String?
    public let date: Date
    public let overallScore: Double
    public let physicalLimitsScore: Double
    public let socialLimitsScore: Double
    public let qualityOfLifeScore: Double
    public let specificSymptomsScore: Double
    public let dizzinessScore: Double
}
```

File: ENGAGEHF/VitalsManager/VitalsManager.swift [hunk 8]
```diff
@@ -145,21 +164,18 @@ public class VitalsManager: Module, EnvironmentAccessible {
...
         let quantityType: HKQuantityType
         
-        switch observationType {
-        case "29463-7": // Weight
+        if observation.code.containsCoding(code: "29463-7", system: FHIRSystem.loinc.url) {
+            // Weight
             hkQuantity = try self.getQuantity(observation: observation)
             quantityType = HKQuantityType(.bodyMass)
-        case "8867-4": // Heart Rate
+        } else if observation.code.containsCoding(code: "8867-4", system: FHIRSystem.loinc.url) {
```
Comment (pauljohanneskraft) [inline ENGAGEHF/VitalsManager/VitalsManager.swift:174]: Can we maybe also put those codes in an enum? They are probably used more than once, e.g. when creating the observations in the first place

File: ENGAGEHF/VitalsManager/VitalsUnit.swift [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+// SPDX-FileCopyrightText: 2024 Stanford University
+//
+// SPDX-License-Identifier: MIT
+//
+
+import Foundation
+import HealthKit
+
+
+enum VitalsUnit: CustomStringConvertible {
```
Comment (pauljohanneskraft) [inline ENGAGEHF/VitalsManager/VitalsUnit.swift:13]: Could we potentially replace this with HKUnit instead? or provide a computed property that returns the HKUnit rather than doing string conversions?

**Review by pauljohanneskraft — CHANGES_REQUESTED** [review]
Very good work! I know it may sound quite negative, but please don't take this in a negative way - just trying to improve the codebase and give suggestions 😊  some things may also simply be personal preference rather than strict requests - feel free to ask me for more extensive feedback if anything is not immediately clear

**Review by PSchmiedmayer — COMMENTED** [review]
Amazing work here @nriedman! Exciting to see this merged and thank you for all the review comments @pauljohanneskraft!

File: ENGAGEHF.xcodeproj/project.pbxproj [hunk 10]
```diff
@@ -1313,6 +1587,14 @@
 				minimumVersion = "3.0.0-beta.1";
 			};
 		};
+		4DC0F6722C34929C0025AB13 /* XCRemoteSwiftPackageReference "ExceptionCatcher" */ = {
+			isa = XCRemoteSwiftPackageReference;
+			repositoryURL = "https://github.com/sindresorhus/ExceptionCatcher.git";
```
Comment (PSchmiedmayer) [inline ENGAGEHF.xcodeproj/project.pbxproj:1592]: We should use a tagged version up to the next major version, e.g., https://github.com/sindresorhus/ExceptionCatcher/releases/tag/v2.0.1

File: ENGAGEHF/Dashboard/Survey/FullSurveyView.swift
Comment (PSchmiedmayer) [inline ENGAGEHF/Dashboard/Survey/FullSurveyView.swift:1]: I think some files here were added despite them to being the focus of this PR, e.g., Survey Elements.

File: ENGAGEHF/Dashboard/Vitals/VitalsCard.swift [hunk 11]
```diff
@@ -15,23 +15,20 @@ struct VitalsCard: View {
     let units: String
     let measurement: (value: String, date: String)?
     
-    @ScaledMetric private var measurementTextSize: CGFloat = 40
+    var measurementTextSize: CGFloat = 40
```
[the remark this answers is not in the document]
Comment (PSchmiedmayer) [reply ENGAGEHF/Dashboard/Vitals/VitalsCard.swift:18]: Agree, this should be a scaled metric.

File: ENGAGEHF/ENGAGEHFStandard.swift [hunk 12]
```diff
@@ -57,13 +57,28 @@ actor ENGAGEHFStandard: Standard, EnvironmentAccessible, OnboardingConstraint, A
...
+        bodyMassMapping?.unit = MappedUnit(
+            hkunit: .gramUnit(with: .kilo),
+            unit: "kg",
+            system: URL(string: "http://unitsofmeasure.org")!,
+            code: "kg"
+        )
+        sampleMapping.quantitySampleMapping[HKQuantityType(.bodyMass)] = bodyMassMapping
+        
+        return sampleMapping
+    }
```
[the remark this answers is not in the document]
Comment (PSchmiedmayer) [reply ENGAGEHF/ENGAGEHFStandard.swift:74]: I think it would be good to by-default store in KG to ideally use SI units where possible. But both the app and the server should understand all units.

File: ENGAGEHF/ENGAGEHFStandard.swift [hunk 13]
```diff
@@ -94,6 +108,16 @@ actor ENGAGEHFStandard: Standard, EnvironmentAccessible, OnboardingConstraint, A
     }
     
     
+    func add(symptomScore: SymptomScore) async {
+        do {
+            let userDoc = try await userDocumentReference
```
Comment (PSchmiedmayer) [inline ENGAGEHF/ENGAGEHFStandard.swift:113]: User Document Reference should be renamed to `patientDocumentReference` and updated in accordance to the changes that @pauljohanneskraft merges today to ideally not break the client.

File: ENGAGEHF/ENGAGEHFStandard.swift [hunk 14]
```diff
@@ -122,15 +146,15 @@ actor ENGAGEHFStandard: Standard, EnvironmentAccessible, OnboardingConstraint, A
         var collectionBucket: String? {
             switch type {
             case HKQuantityType(.bodyMass):
-                return "bodyWeightObservations"
+                return CollectionID.bodyWeightObservations.rawValue
             case HKQuantityType(.bodyMassIndex):
                 return "bodyMassIndexObservations"
```
Comment (PSchmiedmayer) [inline ENGAGEHF/ENGAGEHFStandard.swift:151]: Might make sense to add that to the `CollectionID` type, even if we don't really use that collection everywhere or just return nil here.

File: ENGAGEHF/HeartHealth/HeartHealth.swift [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+    
+    
+    var body: some View {
+        NavigationStack {
+            VStack(alignment: .trailing) {
+                GraphPicker(selection: $vitalSelection)
+                    .padding(.horizontal)
+                VitalsList(vitalSelection: vitalSelection)
+            }
+                .navigationTitle("Heart Health")
```
[the remark this answers is not in the document]
Comment (PSchmiedmayer) [reply ENGAGEHF/HeartHealth/HeartHealth.swift:27]: I think it is fine to keep it like this for now, we could go ahead and improve that with a general localization push when we translate to Spanish, might be a good issue for the future.

File: ENGAGEHF/HeartHealth/SelectionTypes.swift [hunk 16]
```diff
@@ -0,0 +1,175 @@
...
+/// Chosen by GraphPicker in HeartHealth
+enum GraphSelection: CaseIterable, Identifiable, CustomStringConvertible {
+    case symptoms
+    case weight
+    case heartRate
+    case bloodPressure
+    
+    var id: Self { self }
+    
+    var description: String {
```
[the remark this answers is not in the document]
Comment (PSchmiedmayer) [reply ENGAGEHF/HeartHealth/SelectionTypes.swift:23]: I agree with @pauljohanneskraft that we should name these types `localized...` and we should at best return a `LocalizedStringResource`: https://developer.apple.com/documentation/foundation/localizedstringresource. The Apple Documentation provides some good examples on why this is a good tactic. Might be something to address in a smaller issue after that. Applies to all these values here and in other types.

File: ENGAGEHF/HeartHealth/Shared/DescriptionSection.swift [hunk 17]
```diff
@@ -0,0 +1,33 @@
...
+
+struct DescriptionSection: View {
+    var explanationKey: String
+    var quantityName: String
+    
+    
+    var body: some View {
+        Section(
+            content: {
+                Text(LocalizedStringKey(explanationKey))
```
[the remark this answers is not in the document]
Comment (PSchmiedmayer) [reply ENGAGEHF/HeartHealth/Shared/DescriptionSection.swift:20]: Agree, added a comment above. I would suggest to use https://developer.apple.com/documentation/foundation/localizedstringresource.

File: ENGAGEHF/HeartHealth/Shared/GraphSection/DataTypes/AggregatedMeasurement.swift [hunk 18]
```diff
@@ -0,0 +1,20 @@
...
+
+
+/// A graph-ready representation of a vitals measurement
+/// Represents a single, aggregated (averaged across an interval) data point
+struct AggregatedMeasurement: Hashable, Identifiable {
+    var id: Self { self }
+    var date: Date
+    var value: Double
+    var count: Int
+    var series: String
```
Comment (PSchmiedmayer) [inline ENGAGEHF/HeartHealth/Shared/GraphSection/DataTypes/AggregatedMeasurement.swift:19]: In the future we should make this something more type-safe. An issue for the future with different other elements here in this PR.

File: ENGAGEHF/HeartHealth/Shared/GraphSection/VitalsGraphSubViews/Modifiers/VitalsGraph+GestureOverlay.swift [hunk 19]
```diff
@@ -0,0 +1,40 @@
...
+                                    }
+                            )
+                            .gesture(
+                                DragGesture()
+                                    .onChanged { value in
+                                        viewModel.selectPoint(value: value, proxy: proxy, geometry: geometry, clearOnGap: false)
+                                    }
+                            )
+                    }
+                }
```
[the remark this answers is not in the document]
Comment (PSchmiedmayer) [reply ENGAGEHF/HeartHealth/Shared/GraphSection/VitalsGraphSubViews/Modifiers/VitalsGraph+GestureOverlay.swift:37]: Think the idea of a view modifier is fine, in the future we might want to make it a `.gestureOverlay()` overload on a view, as described in https://developer.apple.com/documentation/swiftui/viewmodifier:
```swift
extension View {
    func gestureOverlay(_ viewModel: ViewModel) {
        modifier(GestureOverlay(viewModel: viewModel))
    }
}
```

File: ENGAGEHF/HeartHealth/Symptoms/SymptomsPicker.swift [hunk 20]
```diff
@@ -0,0 +1,53 @@
...
+import SwiftUI
+
+
+struct SymptomsPicker: View {
+    @Binding var symptomsType: SymptomsType
+    
+    
+    var body: some View {
+        Menu(
+            content: {
```
[the remark this answers is not in the document]
Comment (PSchmiedmayer) [reply ENGAGEHF/HeartHealth/Symptoms/SymptomsPicker.swift:18]: I think it is fine to avoid railing closures here, makes it nicer to read what each content is.

File: ENGAGEHF/HeartHealth/Vitals/HKSampleGraph/HKSampleGraph+ViewModel.swift [hunk 21]
```diff
@@ -0,0 +1,115 @@
...
+//
+// SPDX-License-Identifier: MIT
+//
+
+import Foundation
+import HealthKit
+import SpeziViews
+
+
+extension HKSampleGraph {
```
Comment (PSchmiedmayer) [inline ENGAGEHF/HeartHealth/Vitals/HKSampleGraph/HKSampleGraph+ViewModel.swift:14]: Nice job with all the functionality here!

File: ENGAGEHF/HeartHealth/Vitals/HKSampleGraph/HKSampleGraph.swift [hunk 22]
```diff
@@ -0,0 +1,45 @@
...
+//
+// SPDX-License-Identifier: MIT
+//
+
+import HealthKit
+import SpeziViews
+import SwiftUI
+
+
+struct HKSampleGraph: View {
```
Comment (PSchmiedmayer) [inline ENGAGEHF/HeartHealth/Vitals/HKSampleGraph/HKSampleGraph.swift:14]: A very clean API surface!

File: ENGAGEHF/Medications/Medications.swift
Comment (PSchmiedmayer) [inline ENGAGEHF/Medications/Medications.swift:1]: Might not be part of this PR.

File: ENGAGEHF/VitalsManager/VitalsManager.swift [hunk 23]
```diff
@@ -109,17 +119,26 @@ public class VitalsManager: Module, EnvironmentAccessible {
...
+            self.registerSnapshot(
+                collectionReference: userDocRef.collection(CollectionID.kccqResults.rawValue),
+                storage: \.symptomHistory,
+                mapObservation: { $0 }
+            )
+        )
     }
     
-    private func registerSnapshot<T>(
+    private func registerSnapshot<T, V: Decodable>(
```
Comment (PSchmiedmayer) [inline ENGAGEHF/VitalsManager/VitalsManager.swift:138]: Nice job here!

File: ENGAGEHF/VitalsManager/VitalsUnit.swift (hunk 9 above)
Comment (PSchmiedmayer) [reply ENGAGEHF/VitalsManager/VitalsUnit.swift:13]: I think the issue is that HKUnit doesn't do `BPM` and elements that we would need for the visualizations.

**Review by PSchmiedmayer — APPROVED** [review]
Amazing job @nriedman with this PR!! Such a great UI and functionality!

It would be great to address the smaller comments and suggestions and move elements like a reusability, the localization elements, and other larger elements into separate issues.

Last thing that would be important here would be the adaption of the user collection to the patients collection for all the observations to ensure that this works well with the work that @pauljohanneskraft will merge in a few hours 🚀

File: ENGAGEHF.xcodeproj/project.pbxproj (hunk 10 above)
Comment (nriedman) [reply ENGAGEHF.xcodeproj/project.pbxproj:1582]: I actually ended up moving away from initializing HKUnits from a string, which is why this package was included. We can safely remove the dependency entirely.

File: ENGAGEHF/ENGAGEHFStandard.swift (hunk 14 above)
Comment (nriedman) [reply ENGAGEHF/ENGAGEHFStandard.swift:137]: I went ahead and set it to nil, since we set up a listener to every collection in CollectionID. This would be fine if we wanted to listen for BMI and whatnot, but for now I figured we can ignore this case. Definitely me know if it would instead be better to choose which collections to listen to without iterating through every case of CollectionID, however.

**Review by pauljohanneskraft — APPROVED** [review]
